File: python/otel.py
from __future__ import annotations

import threading
from typing import Any

import env

TRACER_NAME = "speaches/realtime"
SERVICE_NAME_DEFAULT = "speaches-plus"

_lock = threading.Lock()
_provider: Any | None = None
_enabled: bool = False

def is_enabled() -> bool:
    return _enabled

def init() -> bool:
    global _provider, _enabled
    with _lock:
        if _provider is not None:
            return _enabled

        endpoint = env.read_str_or_none(env.OTEL_EXPORTER_OTLP_ENDPOINT)
        if not endpoint:
            _enabled = False
            return False

        protocol = env.read_str(
            env.OTEL_EXPORTER_OTLP_PROTOCOL, "http/protobuf"
        ).strip().lower() or "http/protobuf"
        service_name = env.read_str(
            env.OTEL_SERVICE_NAME, SERVICE_NAME_DEFAULT
        ) or SERVICE_NAME_DEFAULT

        try:
            from opentelemetry import trace as ot_trace
            from opentelemetry.sdk.resources import Resource
            from opentelemetry.sdk.trace import TracerProvider
            from opentelemetry.sdk.trace.export import BatchSpanProcessor

            if protocol == "grpc":
                from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
                    OTLPSpanExporter,
                )
                exporter = OTLPSpanExporter(endpoint=endpoint)
            else:
                from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
                    OTLPSpanExporter,
                )
                exporter = OTLPSpanExporter(endpoint=endpoint)

            resource = Resource.create({"service.name": service_name})
            provider = TracerProvider(resource=resource)
            provider.add_span_processor(BatchSpanProcessor(exporter))
            ot_trace.set_tracer_provider(provider)

            _provider = provider
            _enabled = True
            return True
        except ImportError:
            _enabled = False
            return False
        except Exception:
            _enabled = False
            return False

def shutdown() -> None:
    global _provider, _enabled
    with _lock:
        provider = _provider
        _provider = None
        _enabled = False
    if provider is None:
        return
    try:
        flush = getattr(provider, "force_flush", None)
        if callable(flush):
            flush()
        sd = getattr(provider, "shutdown", None)
        if callable(sd):
            sd()
    except Exception:
        pass

def current_span_id_hex() -> str | None:
    if not _enabled:
        return None
    try:
        from opentelemetry import trace as ot_trace

        span = ot_trace.get_current_span()
        ctx = span.get_span_context()
        if not ctx.is_valid:
            return None
        return f"{ctx.span_id:016x}"
    except ImportError:
        return None
    except Exception:
        return None

File: rust/src/otel.rs
#![allow(dead_code)]

use std::sync::OnceLock;

use opentelemetry::trace::TraceContextExt;
use opentelemetry::KeyValue;
use opentelemetry_otlp::{Protocol, WithExportConfig};
use opentelemetry_sdk::trace::SdkTracerProvider;
use opentelemetry_sdk::Resource;
use tracing::Subscriber;
use tracing_opentelemetry::{OpenTelemetryLayer, OpenTelemetrySpanExt};
use tracing_subscriber::registry::LookupSpan;
use tracing_subscriber::Layer;

use super::defaults;

static PROVIDER: OnceLock<SdkTracerProvider> = OnceLock::new();
static OTEL_ENABLED: OnceLock<bool> = OnceLock::new();

pub fn is_enabled() -> bool {
    *OTEL_ENABLED.get().unwrap_or(&false)
}

pub fn current_span_id_hex() -> Option<String> {
    if !is_enabled() {
        return None;
    }
    let span = tracing::Span::current();
    let cx = span.context();
    let span_ref = cx.span();
    let sc = span_ref.span_context();
    if !sc.is_valid() {
        return None;
    }
    Some(format!("{:016x}", sc.span_id()))
}

pub fn try_install_layer<S>() -> anyhow::Result<Option<Box<dyn Layer<S> + Send + Sync + 'static>>>
where
    S: Subscriber + for<'a> LookupSpan<'a> + Send + Sync,
{
    let endpoint = match std::env::var(defaults::env::OTEL_EXPORTER_OTLP_ENDPOINT) {
        Ok(v) if !v.trim().is_empty() => v,
        _ => {
            let _ = OTEL_ENABLED.set(false);
            return Ok(None);
        }
    };

    let protocol = std::env::var(defaults::env::OTEL_EXPORTER_OTLP_PROTOCOL)
        .ok()
        .map(|s| s.trim().to_ascii_lowercase())
        .unwrap_or_else(|| "http/protobuf".into());

    let service_name = std::env::var(defaults::env::OTEL_SERVICE_NAME)
        .unwrap_or_else(|_| defaults::tracing::SERVICE_NAME_DEFAULT.into());

    let resource = Resource::builder()
        .with_attribute(KeyValue::new("service.name", service_name))
        .build();

    let exporter = match protocol.as_str() {
        "grpc" => opentelemetry_otlp::SpanExporter::builder()
            .with_tonic()
            .with_endpoint(endpoint)
            .with_protocol(Protocol::Grpc)
            .build()?,
        _ => opentelemetry_otlp::SpanExporter::builder()
            .with_http()
            .with_endpoint(endpoint)
            .with_protocol(Protocol::HttpBinary)
            .build()?,
    };

    let provider = SdkTracerProvider::builder()
        .with_batch_exporter(exporter)
        .with_resource(resource)
        .build();

    let tracer =
        opentelemetry::trace::TracerProvider::tracer(&provider, defaults::tracing::TRACER_NAME);

    opentelemetry::global::set_tracer_provider(provider.clone());
    let _ = PROVIDER.set(provider);
    let _ = OTEL_ENABLED.set(true);

    let layer = OpenTelemetryLayer::new(tracer);
    Ok(Some(Box::new(layer)))
}

pub fn shutdown() {
    if let Some(provider) = PROVIDER.get() {
        let _ = provider.force_flush();
        let _ = provider.shutdown();
    }
}
